### PR TITLE
chore(ci): add integer layer to code coverage

### PR DESCRIPTION
**Pull request by soonum**

### CODE REVIEW

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

**Comment by soonum** [comment]
@slab-ci code_coverage

File: tfhe/src/integer/server_key/radix/tests.rs [hunk 1]
```diff
@@ -311,14 +336,18 @@ fn integer_unchecked_bitand(param: ClassicPBSParameters) {
 
         let clear_1 = rng.gen::<u64>() % modulus;
 
+        let start = time::Instant::now();
```
Comment (tmontaigu) [inline tfhe/src/integer/server_key/radix/tests.rs:339]: Do we want to keep these timings prints ?

Comment (soonum) [reply tfhe/src/integer/server_key/radix/tests.rs:339]: No you're right. It must be a leftover of my debug phase.

**Comment by IceTDrinker** [comment]
seeing conflicts, might be with my recent removal of an outdated/experimental parameter set

File: tfhe/examples/utilities/generates_test_keys.rs [hunk 2]
```diff
@@ -19,9 +20,10 @@ use tfhe::shortint::parameters::parameters_wopbs_message_carry::{
 };
 use tfhe::shortint::parameters::{
     ClassicPBSParameters, WopbsParameters, ALL_MULTI_BIT_PARAMETER_VEC, ALL_PARAMETER_VEC,
-    PARAM_MESSAGE_1_CARRY_1_KS_PBS, PARAM_MESSAGE_2_CARRY_2_KS_PBS, PARAM_MESSAGE_3_CARRY_3_KS_PBS,
-    PARAM_MESSAGE_4_CARRY_4_KS_PBS,
+    PARAM_MESSAGE_1_CARRY_1_KS_PBS, PARAM_MESSAGE_2_CARRY_2_KS_PBS, PARAM_MESSAGE_2_CARRY_3_KS_PBS,
```
Comment (IceTDrinker) [inline tfhe/examples/utilities/generates_test_keys.rs:23]: we add the 2_3 for which reason ? It seems to me we have coverage params for that ?

File: tfhe/src/integer/public_key/tests.rs [hunk 3]
```diff
@@ -7,12 +7,14 @@ use crate::shortint::ClassicPBSParameters;
 use crate::integer::keycache::KEY_CACHE;
 
 create_parametrized_test!(big_radix_encrypt_decrypt_128_bits {
+    #[cfg(not(feature = "__coverage"))]
```
Comment (IceTDrinker) [inline tfhe/src/integer/public_key/tests.rs:10]: I think we can better use the cfg tarpaulin ?

https://github.com/xd009642/tarpaulin?tab=readme-ov-file#ignoring-code-in-files

Maybe in a follow-up PR to avoid having any more changes on this PR

File: tfhe/src/integer/public_key/tests.rs [hunk 4]
```diff
@@ -49,6 +51,7 @@ fn big_radix_encrypt_decrypt_128_bits(param: ClassicPBSParameters) {
     assert_eq!(clear, dec);
 }
 
+#[cfg(not(feature = "__coverage"))]
 fn radix_encrypt_decrypt_compressed_128_bits(param: ClassicPBSParameters) {
```
Comment (IceTDrinker) [inline tfhe/src/integer/public_key/tests.rs:55]: this is weird ?

File: tfhe/src/integer/server_key/radix/tests.rs [hunk 5]
```diff
@@ -23,15 +37,20 @@ create_parametrized_test!(integer_encrypt_decrypt_256_bits_specific_values);
...
 create_parametrized_test!(integer_smart_add);
+#[cfg(not(feature = "__coverage"))]
 create_parametrized_test! {
     integer_smart_add_128_bits {
         // Skip the 1_1 params for the smart add 128 bits which proved to be the slowest test in our test
         // suite
         PARAM_MESSAGE_2_CARRY_2_KS_PBS,
+        #[cfg(not(feature = "__coverage"))]
         PARAM_MESSAGE_3_CARRY_3_KS_PBS,
+        #[cfg(not(feature = "__coverage"))]
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/tests.rs:48]: not required if the test is marked to not run in coverage

File: tfhe/src/integer/server_key/radix/tests.rs [hunk 6]
```diff
@@ -9,12 +11,24 @@ use crate::integer::server_key::radix_parallel::tests_cases_unsigned::*;
...
+const NB_TESTS: usize = 1;
+#[cfg(feature = "__coverage")]
+const NB_TESTS_SMALLER: usize = 1;
+
+#[cfg(not(feature = "__coverage"))]
 const NB_CTXT: usize = 4;
+#[cfg(feature = "__coverage")]
+const NB_CTXT: usize = 2;
 
 create_parametrized_test!(integer_encrypt_decrypt);
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/tests.rs:33]: shouldn't the macro be updated for coverage (or have a different definition depending on the coverage so that only the coverage parameters are used ?

File: tfhe/src/shortint/parameters/mod.rs [hunk 7]
```diff
@@ -14,6 +14,7 @@ pub use crate::core_crypto::commons::parameters::{
 use crate::core_crypto::prelude::{LweCiphertextListParameters, LweCiphertextParameters};
 use serde::{Deserialize, Serialize};
 
+pub mod coverage_parameters;
```
Comment (IceTDrinker) [inline tfhe/src/shortint/parameters/mod.rs:17]: feature gate/cfg gate for coverage

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs [hunk 8]
```diff
@@ -3911,10 +3923,12 @@ where
     let (cks, sks) = KEY_CACHE.get_from_params(param, IntegerKeyKind::Radix);
     let sks = Arc::new(sks);
 
-    let cks = RadixClientKey::from((cks, NB_CTXT));
+    let nb_ctxt = 4;
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_unsigned.rs:3926]: nb_ctxt = 4.max(NB_CTXT)

maybe ? and indicate we need at least 4 blocks in a comment (I guess that's why this was changed to 4)

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 9]
```diff
@@ -1570,144 +2240,129 @@ macro_rules! define_signed_scalar_comparison_test_functions {
...
-        2,
-        ServerKey::scalar_max_parallelized,
-        std::cmp::max,
-    );
-}
-
-create_parametrized_test!(integer_signed_unchecked_scalar_max_parallelized_128_bits {
+#[cfg(not(feature = "__coverage"))]
+define_signed_scalar_comparison_test_functions!(eq, i128);
+#[cfg(feature = "__coverage")]
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:2292]: no need for feature gate here I think ? worst case make the functions pub if clippy complains they are only enabled for testing and indicate only the i128 are used for regular testing and i8 only are use for coverage

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 10]
```diff
@@ -1512,17 +2052,147 @@ fn test_signed_default_scalar_function<SmartFn, ClearF>(
...
-        let decrypted_result: i128 = cks.decrypt_signed_radix(&encrypted_result);
+        let decrypted_result: T = cks.decrypt_signed_radix(&encrypted_result);
 
         let expected_result = clear_fn(clear_0, clear_1);
         assert_eq!(decrypted_result, expected_result);
     }
 }
 
+#[cfg(not(feature = "__coverage"))]
+fn integer_signed_unchecked_scalar_min_parallelized_i128(
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:2067]: can it be macroized like below ?

do the functions need the gating (if needed we can make them pub to avoid clippy beingannoying)

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 11]
```diff
@@ -1359,97 +1880,110 @@ create_parametrized_test!(integer_is_scalar_out_of_bounds {
...
     let mut rng = rand::thread_rng();
 
-    let num_block = (128f64 / (param.message_modulus.0 as f64).log(2.0)).ceil() as usize;
-
     let (cks, sks) = KEY_CACHE.get_from_params(param, IntegerKeyKind::Radix);
+    let num_block = divide_ceil(T::BITS, param.message_modulus.0.ilog2() as usize);
+    assert_eq!(
+        T::BITS as u32 % param.message_modulus.0.ilog2(),
+        0,
+        "bit width must a multiple of number of bit in a block"
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:1902]: bit width must **be**

the typo is likely everywhere else in the file

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 12]
```diff
@@ -1037,135 +1532,159 @@ macro_rules! define_scalar_comparison_test_functions {
...
+#[cfg(feature = "__coverage")]
+create_parametrized_test!(integer_scalar_max_parallelized_u8 {
+    COVERAGE_PARAM_MESSAGE_2_CARRY_2_KS_PBS,
+});
+
 /// The goal of this function is to ensure that scalar comparisons
 /// work when the scalar type used is either bigger or smaller (in bit size)
 /// compared to the ciphertext
-fn integer_unchecked_scalar_comparisons_edge(param: ClassicPBSParameters) {
+fn test_unchecked_scalar_comparisons_edge(param: ClassicPBSParameters) {
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:1687]: why was this renamed ?

with the create parameterized test below we will have a test called test_test_something ?

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 13]
```diff
@@ -1037,135 +1532,159 @@ macro_rules! define_scalar_comparison_test_functions {
...
-}
-
-fn integer_scalar_max_parallelized_256_bits(params: crate::shortint::ClassicPBSParameters) {
-    test_default_scalar_function(params, 2, ServerKey::scalar_max_parallelized, std::cmp::max);
-}
-
-create_parametrized_test!(integer_unchecked_scalar_min_parallelized_256_bits {
+#[cfg(not(feature = "__coverage"))]
+define_scalar_comparison_test_functions!(eq, U256);
+#[cfg(feature = "__coverage")]
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:1584]: probably does not need the gating, make fn pub if needed to please clippy

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 14]
```diff
@@ -978,17 +1355,135 @@ fn test_default_scalar_function<SmartFn, ClearF>(
...
-        let decrypted_result: U256 = cks.decrypt_radix(&encrypted_result);
+        let decrypted_result: T = cks.decrypt_radix(&encrypted_result);
 
         let expected_result = clear_fn(clear_0, clear_1);
         assert_eq!(decrypted_result, expected_result);
     }
 }
 
+#[cfg(not(feature = "__coverage"))]
+fn integer_unchecked_scalar_min_parallelized_u256(params: crate::shortint::ClassicPBSParameters) {
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:1370]: macroize ? and remove features with pub fn

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 15]
```diff
@@ -737,76 +1017,125 @@ macro_rules! define_signed_comparison_test_functions {
...
-define_signed_comparison_test_functions!(ne);
-define_signed_comparison_test_functions!(lt);
-define_signed_comparison_test_functions!(le);
-define_signed_comparison_test_functions!(gt);
-define_signed_comparison_test_functions!(ge);
-
+#[cfg(not(feature = "__coverage"))]
+define_signed_comparison_test_functions!(eq, i128);
+#[cfg(feature = "__coverage")]
+define_signed_comparison_test_functions!(eq, i8);
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:1099]: remove feature gate, make pub fn where required

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 16]
```diff
@@ -608,56 +812,132 @@ fn test_signed_default_function<SmartFn, ClearF>(
...
 
-        let decrypted_result: i128 = cks.decrypt_signed_radix(&encrypted_result);
+        let decrypted_result: T = cks.decrypt_signed_radix(&encrypted_result);
 
         let expected_result = clear_fn(clear_0, clear_1);
         assert_eq!(decrypted_result, expected_result);
     }
 }
 
+#[cfg(not(feature = "__coverage"))]
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:829]: features gates to remove below using pub fn where appropriate

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 17]
```diff
@@ -358,49 +483,108 @@ use crate::shortint::parameters::{
...
-define_comparison_test_functions!(lt);
-define_comparison_test_functions!(le);
-define_comparison_test_functions!(gt);
-define_comparison_test_functions!(ge);
-
-create_parametrized_test!(integer_unchecked_min_parallelized_256_bits {
+#[cfg(not(feature = "__coverage"))]
+define_comparison_test_functions!(eq, U256);
+#[cfg(feature = "__coverage")]
+define_comparison_test_functions!(eq, u8);
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:489]: feature gates

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 18]
```diff
@@ -181,52 +209,127 @@ fn test_default_function<SmartFn, ClearF>(
...
         params,
         2,
         ServerKey::unchecked_min_parallelized,
-        std::cmp::min,
+        std::cmp::min::<U256>,
+    );
+}
+
+#[cfg(feature = "__coverage")]
+fn integer_unchecked_min_parallelized_u8(params: crate::shortint::ClassicPBSParameters) {
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:237]: feature gates

**Review by IceTDrinker — COMMENTED** [review]
Mainly in the comparison file I think a lot of the feature gates can be removed to simplify the code somewhat, I haven't put it everywhere, let me know if the idea is not clear

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs (hunk 12 above)
Comment (soonum) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:1687]: You're right, it's a mistake.

File: tfhe/src/integer/server_key/radix/tests.rs (hunk 6 above)
Comment (soonum) [reply tfhe/src/integer/server_key/radix/tests.rs:33]: On hold for now.

File: tfhe/src/integer/key_switching_key/test.rs [hunk 19]
```diff
@@ -75,13 +80,17 @@ fn gen_multi_keys_test_crtinteger_to_crtinteger_ci_run_filter() {
...
+    // We retrieve only one keys set from cache since testing key switching with two identical keys
+    // set is meaningless.
     let (client_key_1, server_key_1) =
-        crate::integer::gen_keys_crt(PARAM_MESSAGE_2_CARRY_2_KS_PBS, basis.clone());
+        KEY_CACHE.get_from_params(PARAM_MESSAGE_2_CARRY_2_KS_PBS, IntegerKeyKind::CRT);
+    let client_key_1 = CrtClientKey::from((client_key_1, basis.clone()));
 
-    // We generate a set of client/server keys, using the default parameters:
+    // We retrieve only one keys set from cache since testing key switching with two identical keys
+    // set is meaningless.
```
Comment (IceTDrinker) [inline tfhe/src/integer/key_switching_key/test.rs:90]: I think the comment is slightly wrong we retrieve both from the keycache but they are from different parameter sets

File: tfhe/src/integer/server_key/radix_parallel/tests_unsigned.rs [hunk 20]
```diff
@@ -9,119 +11,224 @@ use super::tests_cases_unsigned::*;
...
+        }
+    }
+);
+create_parametrized_test!(
+    integer_default_rem {
+        coverage => {
+            COVERAGE_PARAM_MESSAGE_2_CARRY_2_KS_PBS,
+            COVERAGE_PARAM_MULTI_BIT_MESSAGE_2_CARRY_2_GROUP_2_KS_PBS,
+        },
+        no_coverage => {
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests_unsigned.rs:176]: I have a horrible doubts, does this work as the parameters may not be imported when we are not in tarpaulin case ?

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests_unsigned.rs:176]: ooooh I think it works thanks to the #[cfg()] in the macro

**Review by IceTDrinker — APPROVED** [review]
I think we are good 🎉 

Thanks a lot for the herculean effort, let's wait to merge so that the 0.5.2 commits will not contain this :)

**Comment by IceTDrinker** [comment]
3_3 params blowing up min/max tests

**Comment by soonum** [comment]
> 3_3 params blowing up min/max tests

What's the parade in this case ? Relaunching and pray ?

**Comment by IceTDrinker** [comment]
> > 3_3 params blowing up min/max tests
> 
> What's the parade in this case ? Relaunching and pray ?

Better, no min/max in 3_3, I don't recall for comparisons @tmontaigu

File: tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs [hunk 21]
```diff
@@ -848,75 +777,88 @@ create_parametrized_test!(integer_signed_min_parallelized_128_bits {
...
 
-    let num_block = (256f64 / (param.message_modulus.0 as f64).log(2.0)).ceil() as usize;
-
     let (cks, sks) = KEY_CACHE.get_from_params(param, IntegerKeyKind::Radix);
+    let num_block = divide_ceil(T::BITS, param.message_modulus.0.ilog2() as usize);
+    assert_eq!(
+        T::BITS as u32 % param.message_modulus.0.ilog2(),
+        0,
+        "bit width must be a multiple of number of bit in a block"
+    );
```
Comment (tmontaigu) [inline tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:799]: This  assert can be removed from unchecked/unchecked_scalar tests, but should be kept in default and smart tests

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:799]: hmm, why ? don't some default or smart ops call unchecked underneath ?

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:809]: It's still the overflow thing, with 3_3 we cannot represent a u8 or u128, or whatever that is a 2^n bits thing we always have more bits.

So in default and smart tests, since we test on non-clean ciphertexts to check that the calls to full_propagate is correctly made, since we actually have more bits the clear 2^n value may overflow while the actual ciphertext won't, so we won't be comparing the same numbers between the cleans and the cipheretxts.

So that assert is there to ensure we don't call these test with block that does not have a number of bits that is a multiple of the total wanted bitsize

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:809]: yes I get that but unchecked could be called on non clean ciphertext ? or does unchecked require clean carries ?

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests_cases_comparisons.rs:809]: it requires clean ciphertext lile almost every non arithmethic operation

**Review by tmontaigu — APPROVED** [review]

**Comment by IceTDrinker** [comment]
still some 3_3 tests

**Review by tmontaigu — APPROVED** [review]

**Comment by IceTDrinker** [comment]
🚀
